File: java/src/test/java/it/giannibombelli/wsc2026/payment/domain/payment/PaymentTransactionDecisionTest.java
package it.giannibombelli.wsc2026.payment.domain.payment;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Una transazione rifiutata rende il pagamento REJECTED; il pagamento diventa ACCEPTED solo se
 * tutte le transazioni accettate rientrano nella finestra di 48h dalla richiesta.
 */
class PaymentTransactionDecisionTest {

    private static final String CLIENT_REFERENCE = "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa";
    private static final Money TOTAL = new Money(new java.math.BigDecimal("100.00"));

    private static final String PAYPAL = "PayPal";
    private static final String KLARNA = "Klarna";

    @Test
    void singleTimelyAcceptLeadsToPaymentAccepted() {
        Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
        Instant providerCompletion = Instant.parse("2026-06-07T11:00:00Z");

        PaymentId pid = new PaymentId(UUID.randomUUID());
        Payment p = Payment.request(pid, new ClientReference(CLIENT_REFERENCE), TOTAL, new Timestamp(requestedAt));
        TransactionId txId = startTransaction(p, PAYPAL, TOTAL, requestedAt.plusSeconds(1));

        PaymentResultEvents evt = p.acceptTransaction(txId, new Timestamp(providerCompletion));

        assertThat(evt).isInstanceOf(PaymentResultEvents.PaymentAccepted.class);
        assertThat(p.status()).isEqualTo(PaymentStatus.ACCEPTED);
    }

    @Test
    void anySingleRejectRejectsTheEntirePayment() {
        Instant requestedAt = Instant.parse("2026-06-07T09:00:00Z");

        PaymentId pid = new PaymentId(UUID.randomUUID());
        Payment p = Payment.request(pid, new ClientReference(CLIENT_REFERENCE), TOTAL, new Timestamp(requestedAt));
        TransactionId txA = startTransaction(p, PAYPAL, new Money(new java.math.BigDecimal("40.00")), requestedAt.plusSeconds(1));
        p.acceptTransaction(txA, new Timestamp(requestedAt.plusSeconds(3600)));

        TransactionId txB = startTransaction(p, KLARNA, new Money(new java.math.BigDecimal("60.00")), requestedAt.plusSeconds(2));
        PaymentResultEvents evt = p.rejectTransaction(txB, new Description("declined"));

        assertThat(evt).isInstanceOf(PaymentResultEvents.PaymentRejected.class);
        assertThat(p.status()).isEqualTo(PaymentStatus.REJECTED);
    }

    @Test
    void allAcceptedButInsufficientCoverageStaysProcessingNotAccepted() {
        Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
        Instant txTime = Instant.parse("2026-06-07T10:30:00Z");

        PaymentId pid = new PaymentId(UUID.randomUUID());
        Payment p = Payment.request(pid, new ClientReference(CLIENT_REFERENCE), TOTAL, new Timestamp(requestedAt));
        TransactionId txId = startTransaction(p, PAYPAL, new Money(new java.math.BigDecimal("60.00")), requestedAt.plusSeconds(1));

        p.acceptTransaction(txId, new Timestamp(txTime));

        assertThat(p.status()).isEqualTo(PaymentStatus.PROCESSING);
    }

    @Test
    void paymentAcceptedOnlyWhenAllTxAcceptedAndAllTimesWithin48hWindow() {
        Instant requestedAt = Instant.parse("2026-06-07T08:00:00Z");
        Instant tx1 = Instant.parse("2026-06-07T09:00:00Z");
        Instant tx2 = Instant.parse("2026-06-08T06:59:59Z");

        PaymentId pid = new PaymentId(UUID.randomUUID());
        Payment p = Payment.request(pid, new ClientReference(CLIENT_REFERENCE), TOTAL, new Timestamp(requestedAt));
        TransactionId txA = startTransaction(p, PAYPAL, new Money(new java.math.BigDecimal("60.00")), requestedAt.plusSeconds(1));
        p.acceptTransaction(txA, new Timestamp(tx1));
        TransactionId txB = startTransaction(p, KLARNA, new Money(new java.math.BigDecimal("40.00")), requestedAt.plusSeconds(2));
        PaymentResultEvents finalEvt = p.acceptTransaction(txB, new Timestamp(tx2));

        assertThat(finalEvt).isInstanceOf(PaymentResultEvents.PaymentAccepted.class);
        assertThat(p.status()).isEqualTo(PaymentStatus.ACCEPTED);
    }

    @Test
    void providerTimeBeyond48hDoesNotCountAsTimelyAcceptForFinalState() {
        Instant requestedAt = Instant.parse("2026-06-01T12:00:00Z");
        Instant timely = Instant.parse("2026-06-01T13:00:00Z");
        Instant late = Instant.parse("2026-06-06T00:00:00Z");

        PaymentId pid = new PaymentId(UUID.randomUUID());
        Payment p = Payment.request(pid, new ClientReference(CLIENT_REFERENCE), TOTAL, new Timestamp(requestedAt));
        TransactionId txA = startTransaction(p, PAYPAL, new Money(new java.math.BigDecimal("50.00")), requestedAt.plusSeconds(1));
        p.acceptTransaction(txA, new Timestamp(timely));
        TransactionId txB = startTransaction(p, KLARNA, new Money(new java.math.BigDecimal("50.00")), requestedAt.plusSeconds(2));

        assertThatThrownBy(() ->
            p.acceptTransaction(txB, new Timestamp(late))
        ).isInstanceOf(IllegalStateException.class);

        assertThat(p.status()).isEqualTo(PaymentStatus.PROCESSING);
    }

    private TransactionId startTransaction(Payment payment, String provider, Money amount, Instant startedAt) {
        TransactionId transactionId = EntityId.generate(TransactionId::new);
        payment.startTransaction(
            transactionId,
            Provider.fromLabel(provider),
            new ProviderReference(UUID.randomUUID()),
            amount,
            new Timestamp(startedAt)
        );
        return transactionId;
    }
}

File: typescript/tests/payment/domain/payment/paymentTransactionDecision.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { providerFromLabel } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';


const CLIENT_REFERENCE = 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa';
const TOTAL = new Money(100);
const PAYPAL = 'PayPal';
const KLARNA = 'Klarna';

describe('PaymentTransactionDecision', () => {
  it('single timely accept leads to payment accepted', () => {
    const requestedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));
    const providerCompletion = new Timestamp(new Date('2026-06-07T11:00:00.000Z'));

    const pid = new PaymentId(Uuid.generate());
    const p = Payment.request(pid, new ClientReference(Uuid.fromString(CLIENT_REFERENCE)), TOTAL, requestedAt);
    const txId = startTransaction(p, PAYPAL, TOTAL, requestedAt.plusSeconds(1));

    const evt = p.acceptTransaction(txId, providerCompletion);

    expect(evt!.kind).toBe('PaymentAccepted');
    expect(p.status()).toEqual(PaymentStatus.ACCEPTED);
  });

  it('any single reject rejects the entire payment', () => {
    const requestedAt = new Timestamp(new Date('2026-06-07T09:00:00.000Z'));

    const pid = new PaymentId(Uuid.generate());
    const p = Payment.request(pid, new ClientReference(Uuid.fromString(CLIENT_REFERENCE)), TOTAL, requestedAt);
    const txA = startTransaction(p, PAYPAL, new Money(40), requestedAt.plusSeconds(1));
    p.acceptTransaction(txA, requestedAt.plusSeconds(3600));

    const txB = startTransaction(p, KLARNA, new Money(60), requestedAt.plusSeconds(2));
    const evt = p.rejectTransaction(txB, new Description('declined'));

    expect(evt.kind).toBe('PaymentRejected');
    expect(p.status()).toEqual(PaymentStatus.REJECTED);
  });

  it('all accepted but insufficient coverage stays processing not accepted', () => {
    const requestedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));
    const txTime = new Timestamp(new Date('2026-06-07T10:30:00.000Z'));

    const pid = new PaymentId(Uuid.generate());
    const p = Payment.request(pid, new ClientReference(Uuid.fromString(CLIENT_REFERENCE)), TOTAL, requestedAt);
    const txId = startTransaction(p, PAYPAL, new Money(60), requestedAt.plusSeconds(1));

    p.acceptTransaction(txId, txTime);

    expect(p.status()).toEqual(PaymentStatus.PROCESSING);
  });

  it('payment accepted only when all tx accepted and all times within 48h window', () => {
    const requestedAt = new Timestamp(new Date('2026-06-07T08:00:00.000Z'));
    const tx1 = new Timestamp(new Date('2026-06-07T09:00:00.000Z'));
    const tx2 = new Timestamp(new Date('2026-06-08T06:59:59.000Z'));

    const pid = new PaymentId(Uuid.generate());
    const p = Payment.request(pid, new ClientReference(Uuid.fromString(CLIENT_REFERENCE)), TOTAL, requestedAt);
    const txA = startTransaction(p, PAYPAL, new Money(60), requestedAt.plusSeconds(1));
    p.acceptTransaction(txA, tx1);
    const txB = startTransaction(p, KLARNA, new Money(40), requestedAt.plusSeconds(2));
    const finalEvt = p.acceptTransaction(txB, tx2);

    expect(finalEvt!.kind).toBe('PaymentAccepted');
    expect(p.status()).toEqual(PaymentStatus.ACCEPTED);
  });

  it('provider time beyond 48h does not count as timely accept for final state', () => {
    const requestedAt = new Timestamp(new Date('2026-06-01T12:00:00.000Z'));
    const timely = new Timestamp(new Date('2026-06-01T13:00:00.000Z'));
    const late = new Timestamp(new Date('2026-06-06T00:00:00.000Z'));

    const pid = new PaymentId(Uuid.generate());
    const p = Payment.request(pid, new ClientReference(Uuid.fromString(CLIENT_REFERENCE)), TOTAL, requestedAt);
    const txA = startTransaction(p, PAYPAL, new Money(50), requestedAt.plusSeconds(1));
    p.acceptTransaction(txA, timely);
    const txB = startTransaction(p, KLARNA, new Money(50), requestedAt.plusSeconds(2));

    expect(() => p.acceptTransaction(txB, late)).toThrow();

    expect(p.status()).toEqual(PaymentStatus.PROCESSING);
  });

  function startTransaction(payment: Payment, provider: string, amount: Money, startedAt: Timestamp): TransactionId {
    const transactionId = generateId((value) => new TransactionId(value));
    payment.startTransaction(
      transactionId,
      providerFromLabel(provider),
      new ProviderReference(Uuid.generate()),
      amount,
      startedAt
    );
    return transactionId;
  }
});
